### DISCUSSION TITLE
Polrfire SoC Temperature driver

### DISCUSSION THREAD

**Discussion by MarkAtMicrochip** [polarfire-soc/polarfire-soc-documentation#494] [Linux drivers for PolarFire Soc]
Is there a driver for reading the on-die temperature sensor in Linux? I can't seem to find what I'm looking for in the documentation. I know that the TVS (Temperature and Voltage Sensor) block holds the 16-bit temperature value. However, I don't know whether there's a way for the MSS to read it and whether we have it enabled in our Linux build. Thanks!

**Comment by hughbreslin** [comment]
Hey @MarkAtMicrochip :) theres no built in driver for doing this unfortunately. There are two ways you can access the TVS, either using devmem and reading via the SCB (system controller bus), you can find the registers described in the PolarFire SoC Register map, in MPFS Instance Map -> PFSOC_CONTROL_SCB:

![image](https://github.com/user-attachments/assets/2b5475e0-9259-4ac7-b209-1fac0e3d31ae)

Note the addresses shown here should have 0x3000_0000 added on, e.g TVS_CONTROL is at 0x37020008 not 0x07020008, you need the 0x3... to access the SCB. 

The alternative option is to use the TVS fabric macro (it appears as an IP core in your design) and connect it via a FIC to read the values. Personally I prefer this approach as you can use the core with other parts of your design and raise interrupts from it etc. If you would like to go this route please open a tech support case and we could provide you with an APB wrapper for the IP core to connect it up easier.

**Reply by MarkAtMicrochip** [reply]
Thanks Hugh!

**Reply by msdarici** [reply]
when I try to access register I get an error :
root@icicle-kit-es-amp:~# devmem2 0x201220a4 w 0x10000
/dev/mem opened.
Error at line 90, file devmem2.c (1) [Operation not permitted]

This codes worked here : https://www.youtube.com/watch?v=88xk1dV3jDs&list=PL9B4edd-p2agS6QkXUZjAvjlPiyAO5RP0&index=7&ab_channel=MicrochipTechnology%2CInc.

My latest Image was Linux + FreeRTOS amp demo image. Actually I try to access temperature value;

root@icicle-kit-es-amp:~# devmem2 0x3702000C 
/dev/mem opened.
Error at line 90, file devmem2.c (1) [Operation not permitted]

Same error.
How can I solve?
Relative links : https://microchip.my.site.com/s/article/Accessing-TVS--Temperature-Voltage-Sensor--Registers-in-PFSoC

**Comment by zmcginty-thinkOrbital** [comment]
you need to change this setting in the kernel config; CONFIG_IO_STRICT_DEVMEM=n
If you're building linux in buildroot; do this through linux kernel menuconfig, or if you're building with yocto, meta-mchp-polarfire-soc/meta-mchp-polarfire-soc-bsp/recipes-kernel/linux/files/mpfs-disco-kit/mpfs_cmdline.cfg for the discovery kit
